### DISCUSSION TITLE
Import calibration into myPTV

### DISCUSSION THREAD

**Discussion by jennifer-yin** [ronshnapp/MyPTV#38] [Q&A]
Hello! I am trying to transition my PTV process (originally matlab/c++) to python by adopting MyPTV. My camera system is currently calibrated using a 2D calibration target, and I have a good calibration in terms of transformation and rotation matrices from image to world and world to image. I am trying to translate this into the calibration format that MyPTV expects, but have not been successful recreating the calibration in MyPTV. Currently, for O, I am using the vectors pointing from world origin to the camera, and Euler angles to transform world coordinates to image coordinates.  I have tried other options too (ie. camera to world transformations), but the calibration gui are vastly different from what I have. Any idea on why this might be, how to fix it, or other creative ways to address this? 

Thanks! 
Jenn

**Comment by ronshnapp** [comment]
Hi Jenn,

The calibration in MyPTV is as you wrote - the O is the position of the camera in the lab coordinates and the omega is the vector of Euler angles for the camera rotation (in radians). There are two possible reasons why you're havin a bit of difficulties:
1) In MyPTV the "y" axis of the cameras points down. Therefore, we commonly need to add a rotation of \pi radians around the camera's z axis. See, for example the example folder that we have in the repository. 
2) Another possible issue is the need to specify the focal length (f) in pixel sizes. The focal length is described in Section 2.2.4 of the manual. 
Try looking at these factors to see if this solves the issues. If not, you could send me your calibration images and I could try to figure this out.

Best,
Ron

**Comment by jennifer-yin** [comment]
Hi Ron,
[calibration-images.zip](https://github.com/ronshnapp/MyPTV/files/11805665/calibration-images.zip)


I wasn't able to figure this out, so I'm uploading my calibration images. The pixel size is 0.0048mm and focal length of the lens is 12mm. The calibration txt file includes the calibration parameters of the calibration method I am currently using. The Tinv and Rinv represent translation and rotation from world to image coordinates. In the txt file, camera 0 corresponds with cam1 image, camera 1 to cam2 image and so forth. On the calibration slate, there are 2 lines which represent the x and y axis, starting from the world origin (0,0,0). I'd love to hear your thoughts on these images and if you think it would be possible to use them to generate a calibration in MyPTV. 

Thanks! 
Jenn

**Reply by ronshnapp** [reply]
Thanks @jennifer-yin. I will look at your images and all through the weekend and let you know probably by Sunday.

Ron

**Comment by ronshnapp** [comment]
Hi Jenn,

I've had a look at the images you sent. I didn't have the target file for your target so I couldn't cimplete a full calibration for your system. Also, there are a few issues that I describe below. Nevertheless, I could still roughly estimate what the Euler angles of the cameras assuming that the coordinate system you're using is as I drew on the image below. 
In this case, I believe a good initial guess for the Euler angles would be something like:

cam1: $\theta_x$ = $3/4 \pi$ ; $\theta_y$ = 0 ; $\theta_z$ = 0
cam2: $\theta_x$ = $3/4 \pi$ ; $\theta_y$ = $1/4 \pi$ ; $\theta_z$ = 0 
cam3: $\theta_x$ = $3/4 \pi$ ; $\theta_y$ = $-1/4 \pi$ ; $\theta_z$ = 0 
cam4: $\theta_x$ = $5/8 \pi$ ; $\theta_y$ = 0 ; $\theta_z$ = 0 

The reason I used x rotations in all of the cameras is because then I get the origin of the camera space coordinates at the top left corner of the image with the "y" axis pointing down, which is how the camera pixel coordinates work in MyPTV (see image below). Also, based on the focal length and camera pixel size, the f initial guess should be around 12/0.0048 = 2500. 
Try it out and see if this works for you.

As for the other issues you might want to take care of:
1) The biggest problem is that the calibration target has points only on a single plane. To really make sure that the calibration works well, we need to have the calibration points on at least two different planes (and the more the better). See for example the picture of a calibration target I used in the past that has three planes (i.e. points with three different "z" values).
2) only camera 1 sees all of the calibration points, whereas the other cameras see only certain parts of the calibration target. While this is not critical, it would definitely lead to inferior calibration, and it might also lead to some difficulties when you need to pair the calibration points with the data from the target file. 
3) Many of the calibration points in the images are blurred because they are out of focus. I would suggest trying to fix this, maybe by closing the iris of the camera lenses and using a stronger illumination.
 
Let me know how it all goes. Good luck!

Best,
Ron

![coords](https://github.com/ronshnapp/MyPTV/assets/34024910/bc58e91b-cdc9-4c8d-89e9-72bcb936301c)
![cal_image](https://github.com/ronshnapp/MyPTV/assets/34024910/5f4d9d3f-db78-4050-8a25-869b5c18a192)
